### ISSUE TITLE
can `static T` be removed from the language?

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#10085]
I'm skeptical but opening this as a centralized place to discuss this since @araq mentioned the following:

https://github.com/nim-lang/Nim/pull/10080#issuecomment-449656127
> static T needs to be removed from the language IMHO. Just allow to pass integer values as type parameters turning the value N into range[0..N-1] would allow almost the same patterns without this never ending clusterfuck of bugs, edge cases and downright unspecified behaviour.

* how would this work without `static T` ?
```nim
macro foo(a: int, b: static int): untyped # a becomes NimNode, not b
```

**Comment by GULPF** [comment]
One nice use of static[T] is compile time validation. This is used to validate time format strings in the times module for example. There are lots of bugs with static[T], but it's a really cool feature and I would be sad to see it removed.

**Comment by mratsim** [comment]
Removing `static[T]` would be deal breaker for me.

I need an alternative that can be passed to proc at compile-time and enforce static evaluation.

One alternative that I like is the notion of [`PhantomType`](https://doc.rust-lang.org/rust-by-example/generics/phantom.html) in Rust, if they could be generalized to values.

> A phantom type parameter is one that doesn't show up at runtime, but is checked statically (and only) at compile time.
> Data types can use extra generic type parameters to act as markers or to perform type checking at compile time. These extra parameters hold no storage values, and have no runtime behavior.

Basically allow this:

```Nim
type Foo[T] = object
  N: static int # Phantom type
  x: seq[T]
```

We would remove all static semantic checking from the compiler (except for arrays). However those static values are still carried around at compile-time and users can do their own checks on them.

In time I expect a `static` library to rise, maybe in the spirit of [`nim-contracts` ](https://github.com/Udiknedormin/NimContracts) but for compile-time contracts checking.

**Comment by Araq** [comment]
> ``macro foo(a: int, b: static int): untyped``

That's exactly what's wrong. ``3`` is not of type ``static int`` in Nim, it is of type ``int`` (check the manual if you don't believe me). However, 3 matches in overloading resolution the type ``static int`` even better than it does match ``int``. There is a very weird type relation going on here, comparable to C++'s ``const`` fiasco. We copied C++'s design mistake and I've never read any good justification for it. Furthermore before ``static T`` was added to Nim, Nim already had the capabilities to turn integer values into types via the ``range`` construct.

**Comment by cooldome** [comment]
I am using `static T` now and would like to use it even more.
Ideally I would like to define FX rate like this:
```
type
  Currency = enum
     USD, EUR, GBP, JPY, ...

  FXRate[from_ccy: static[Currency], to_ccy: static[Currency], date: static[Date]] = distinct float
```

IMO, feature is really needed. Implementation details can be reconsidered if there are better ideas, but the feature needs to stay.

As side note, it is really discouraging and unfair discussion from point of view of the original author and the original author deserves some respect for his significant contribution.

New idea needs to be discussed, but no one should claim that his new idea is going to replace/remove existing implementation. Only when idea is implemented and can be tried by users, we can compare implementations then we can discuss the potential removal of old implementation. Therefore we compare implementation to implementation and we don't compare idea to the implementation. Existing implementation is by default better than any idea until new idea gets implementation in the real code.

**Comment by cooldome** [comment]
As an idea how ``static[T]`` and ``range[T]`` can be implemented. What type  zero `0` has? One can think it is `int`, but zero actually has infinitely many types: `int`, `int32`, `range[-1..1]`, `range[-1..2]`, `range[-2..3]` and so on. Then isn't it better to claim that any `PNode` constant expression is a type of its own and then have convertible relation to `int`, `int32` and so on. literal `PNode` is a type.
.

**Comment by andreaferretti** [comment]
I have been hoping for static support to improve for a while, it would be sad to see it go away. By the way, it was the feature that first convinced me to use nim, even with all its bugs

**Comment by timotheecour** [comment]
/cc @Araq @mratsim @andreaferretti 

it seems a lot of complication/bugs in code is due to fact that a static param of type `T` will have a type with a different AST, namely `static[T]` (eg: see code inside explicitGenericSym, near ` arg = newTypeS(tyStatic, c)`)

what if instead of modifying the AST of the type (by wrapping type `T` inside `static[T]`) we did this instead:
* add a new enum member `sfStatic` in `TSymFlag`
so a `sym: TSym` of kind `skParam` or `skGenericParam` (ie a symbol corresponding to a generic or normal parameter) would have:
```
sym.flags.incl sfStatic
```
and its type left unwrapped, it `int` instead of `static[int]` in the internal AST representation

This would eliminate the need for all those `skipTypes` on `tyStatic`, ie, it wouldn't result in special cases all over the place

**Comment by dom96** [comment]
I am behind @Araq here. If we can simplify the language by removing this feature and replacing it ,with something that exists already which allows the same semantics, then we should do it.

That said, @Araq seems to imply that `range` can be used in `static[T]`s place, but can it really?

**Comment by andreaferretti** [comment]
No, it can't. That can only work for static[int], and even then only for that particular semantic. It can't work for other use cases for static ints (for instances tensors and matrices of known dimensions, integers modulo N...), and of course it can't replace the use of static types other than int

**Comment by zah** [comment]
So, Christmas is the time to spread heresy :)

There are quite a few claims in this thread that need debunking:

**1. `static T` can be replaced with `range` types**

This covers only the "array size" use case. Static values have many other use cases. I've already written code where macros are being given static values of user-defined object types or function pointers, which are not easy to handle under any of the simplistic alternatives proposed so far (i.e. the macro cannot just inspect the AST and use accessors such as `intVal` to obtain a function pointer to call).

**2. `static T` is weird as it copies C++'s const**

I've always stated that `static` is no different than `var`. Both modifiers are used to create a subset of their respective base type which justifies their higher precedence is overload resolution.

**3. `static T` is too buggy, nobody knows how to fix it**

I remember fixing more than 10 `static T` issues in a [single day][1] once. There is nothing wrong with the design and many of the issues are not that hard to fix. The fact that I can't dedicate a lot of time to working on Nim in the last few months should not inform language design decisions.

[1]: https://github.com/nim-lang/Nim/commits?after=24966e006a56e25fc1cb5c03d1d8ac12dc69d557+1&author=zah

Most of the time, the `static T` issues fall into one following categories:

**A) missing `skipTypes`**

This is code spread around the compiler that is not aware of the possible existence of `tyStatic` as a modifier. These bugs are usually very easy to fix and they just need patience and willingness to investigate. My goal had always been to reach a future where Nim implicitly computes everything that can be evaluated at compile-time and having `tyStatic` in the type system still seems the most reasonable way to represent that (it's less appropriate to attach `sfStatic` flags to symbols, because anonymous expressions can be statically evaluated too). Many more places in the compiler will have to learn to deal with `tyStatic` types before this vision can be realized.

**B) too eager evaluation of generic parameters**

These are issues that are usually triggered in code like this:

```nim
type Foo[N: static int] = array[bar(N) + 1, when N > 0: int else: float]
```

Now, such issues can be quite painful to get right, but guess what - implementing a generic type system is hard. Whatever you replace `static T` with, you'll get the same inherent complexity staring back at you. I can replicate many of the `static T` issues working only with regular generic params and `when` statements based on the `is` operator.

**C) Faulty behaviour of the `array` type.**

The array type is quite special because its canonical representation uses a `range` type to specify its length, but we also allow the `static int` alternative treatment. This has been a source of additional complexity and bugs over the years, which could have been avoided if we hadn't had this duality in the design. Personally, I would prefer to leave only the `static int` treatment as this is what most people expect.

--------------------------------------

With all that said, I'm supportive and I hope to implement the following reform of `static` parameters in macros:

**Rationale:** 

It's more consistent/beneficial if all macro parameters are always AST. This allows you to more easily insert them in the produced AST results and allows you to inspect the precise syntax used at the call-site.

**Proposed reform:** 

1) `static Foo` in macros is typed as a `NimNode` (with some caveats, see point 3). 
2) The AST tree matches precisely what was written at the call-site
3) There is a new magic called `staticValue` (which is similar to `getType`) that returns the well-typed static value the input expression evaluates to. This will take into account the `static T` type of the parameter and will be resolved to the same type.

The same treatment should be applied to `typedesc` parameters to reach full consistency (`staticValue` plays quite a similar role to `getType`).

**Comment by deansher** [comment]
With respect to types, type-level computation, and compile-time computation in general, it seems to me that Nim is within easy reach of a wonderful point in the programming language design spectrum. I picture it as capturing what's best about [Idris](https://www.idris-lang.org/) in these areas, except without crossing the line into dependent typing, which adds a lot of complexity. `static[T]` feels to me so far like evidence that we haven't quite hit this design point yet.

Here is the design point I am hoping Nim will reach:

- The programmer can write Nim code that runs at compile time.
- Types are first-class values at compile time: they can be passed around, stored, and manipulated like any other values.
- Types can only be computed at compile time. Also they can only be examined at compile time, although this is often not apparent because expressions that examine types are evaluated at compile time into constants.
- Types can have ("generic") parameters that are arbitrary Nim data types.
- Much type checking is deferred until after the programmer's compile-time code has been executed, in which case it can use types computed at compile time.
- There is a bare minimum of type inference: just enough inference to be reasonably convenient, but following simple enough rules for the programmer to readily understand and predict type inference and thus its effect on run-time behavior (notably through overloading).

Here's why hitting roughly this design point is important to me:

- Static types are painfully constraining unless they can be at least parameterized, and ideally computed. For example, functional programming in Go is pretty much a non-starter because its type system is strict enough that functional idioms like `map` would have to be explicitly typed, but not parameterizable enough to allow them.
- As programmers strive to capture more of their invariants in the type system, they push even fairly complex type systems (such as Haskell) to steadily increase in complexity. The type system becomes a language of its own -- sometimes a fiendishly complex language. (See, for example, the [Glasgow Haskell Compiler's numerous language extensions](https://downloads.haskell.org/~ghc/8.2.1/docs/html/users_guide/glasgow_exts.html).)
- Especially when a complex type system is combined with complex type inference, it becomes extraordinarily subtle and difficult for the programmer to know what types will be inferred and thus (notably in the face of overloading) what the program will do.
- By using the same language for compile-time and run-time computation, I see an opportunity to allow programmers to flexibly and naturally express whatever invariants they choose to enforce in the type system.

Idris has most of the above advantages, but it goes further to hit a more aggressive design point: it largely creates an illusion that compile-time calculation and run-time calculation can be freely intermixed. It does this by extending type inference into a sophisticated proof engine and proving that type invariants hold even when compile-time and run-time calculation appear to be mixed. Unfortunately, this requires substantial awareness and support from the programmer. This is surely worth exploring, but I am skeptical about its being a good practical solution. (If you want to learn more, you could try [a tutorial](http://idris.live/understanding_dependent_types) that I wrote a couple of years ago when I was trying to get my brain around Idris.)

`static[T]` seems to me so far (as a Nim newbie) as an awkward way to draw a line between compile-time computation and run-time computation. I don't think `static[T]` is a different type than `T` -- not even analogously to how `var T` is a different type than `T`. Or, to come at this from a different direction, I don't think it is productive to say "we do our compile-time computation in Nim itself, except that every value involved in that computation has a different type than if the same code ran at run-time".

I'll come at this from yet another direction: if we are using the same language for compile-time computation, then our compile-time code is statically typed, too! In the compile-time execution environment, does `static[T]` mean "one extra meta-level of static from the rest of this code"? In a proc marked `{.compileTime.}`, does `let x: int = 3` implicitly mean `let x: static[int] = 3`? But then what about a proc not so marked that I choose to use at compile time?

I don't have a proposal yet. And even if I did, I think combining it with all this conceptual material would be poor chunking. So I'm going to let this comment fly and then start thinking about a proposal.

**Comment by timotheecour** [comment]
@zah I agree with some of your points (esp 1 and 3C) but I still stand by my proposal from https://github.com/nim-lang/Nim/issues/10085#issuecomment-449839318 (which doesn't modify AST, and avoids special casing for static everywhere in compiler).
Your only reference to it is:
> it's less appropriate to attach sfStatic flags to symbols, because anonymous expressions can be statically evaluated too

can you provide a concrete example where `sfStatic` wouldn't work but current approach would work?

> My goal had always been to reach a future where Nim implicitly computes everything that can be evaluated at compile-time and having tyStatic in the type system still seems the most reasonable way to represent that

Isn't that already the case?
You can overload by `static` vs `non-static` and Nim already tries to evaluate at compile time except when explicitly given a `var` or `let` input:
```
proc bar[T](a:T): auto = a
proc fun(a: static int) = echo "ok1"
proc fun(a: int) = echo "ok2"
let a = 3
fun(a)
fun(3)
const a2 = 3
fun(a2)
fun(bar(a2*a2))
```
prints:
```
ok2
ok1
ok1
ok1
```

**Comment by zah** [comment]
> can you provide a concrete example where sfStatic wouldn't work but current approach would work?

```nim
takesStaticInt(binom(5, 3))
```

In the sem pass, we prefer to the keep the AST close to its original form. The `binom` call here will be represented as `nkCall` during signature matching and there won't be a symbol to which to attach a `sfStatic` flag. It's possible to add `evalutedValue` field to `PNode` itself, but this seems more excessive and requires all type matching functions in the compiler to work with AST nodes instead of just types.

> Isn't that already the case?
> You can overload by static vs non-static and Nim already tries to evaluate at compile time except when explicitly given a var or let input:

This is true from the perspective of the user, but the internal on-demand evaluation of static parameters is one of the reasons @Araq is so vocal against them. In the ideal implementation, we will implicitly perform more aggressive constant folding everywhere where it's possible and we will store the evaluated values in the type system (in the `tyStatic`'s `n` field). Then, there woudn't be any need for on-demand `semConstExpr` in `typeRel` and friends.

**Comment by timotheecour** [comment]
@zah
```nim
proc binom(a, b: int): int = a+b
proc takesStaticInt(foo: static int) = discard
takesStaticInt(binom(5, 3))
```

> there won't be a symbol to which to attach a sfStatic flag

that would be `foo` ?

**Comment by zah** [comment]
Sure, you can attach a value to the param symbol after the overload resolution is complete and you have selected the callee, but you need to be able to reason about static values earlier - while the overload resolution is underway.

One alternative refactoring that I have considered is introducing a `PNode` field called `originalAst` (or `rewrittenAs`). It would allow us to perform arbitrary rewrites in the sem pass without destroying information. There are few other rewrites in the sem pass besides static evaluation that can use this field, but introducing it will create a whole new set of problems (missing `skipNodes` calls). Also, most compiler textbooks will tell you that AST nodes do not belong in the type system. The rationale for this is simple - there are "only" 64 types at the moment, but hundreds of AST node kinds that can muddle the water some more.

**Comment by zah** [comment]
> Seriously, most of the time I don't know how to treat it internally in the compiler. It is that bad. ;-)

Since the problem is that @Araq claims that he is not able to reason about static types, I can spend some time writing a more detailed internal documentation explaining the rationale and the mechanics of all of the types related to `tyStatic` with their invariants and possible states. Some of the details are ugly (`tyFromExpr`), but quite necessary. My impression is that Araq is a bit prejudiced and refuses to dive deeper into the current implementation.

**Comment by andreaferretti** [comment]
Seems quite natural to me: in the static context, x is just a variable, you cannot use it for the size of an array. The error is the same you would get without a static block

**Comment by Clyybber** [comment]
@deansher What else would you expect it to do?

**Comment by Clyybber** [comment]
@deansher It works in your second example, because the const block introduces yet another static level. If you for example move the `let y: int = 2` out of the const block, you will again get the error that `y` is not available at compile time. This is because a const is essentially a static that is available at runtime too.
So in your example we have essentially three layers:
Runtime: which is empty in your example
Compile time: in which the `static:` block resides in
"Meta-compile" time: in the const block.
I agree that the error message or in general the concept of meta static levels is confusing, but it works as expected AFAICT.

**Comment by zah** [comment]
With implicit static evaluation, it's possible to design a language with a much more naturally looking mixture of compile-time and run-time code. Some var and let bindings may be implicitly considered as constants when their value is known at a particular location in the code and some loops may be implicitly unrolled when they iterate over compile-time data (i.e. similar to how the `fields` iterator in unrolled).

I think Araq would prefer if Nim continues to use the current more explicit constructs for compile-time/run-time discrimination as they are certainly able to express anything possible under a more implicit scheme and their only downside is that a bit more typing is required sometimes (e.g. to unroll a loop over a compile-time value, you may need to define and call a macro).

The added value of the explicit approach is that you clearly see the boundaries of the execution phases in the code, but I think that semantic highlighting in the IDE can be a viable alternative solution for this even under a more implicit approach.

**Comment by deansher** [comment]
Thanks for the perspective! I feel as though we are gradually moving toward something that I, at least, think is the essential next step: a clear conceptual framework and a consistent way to talk about it in English. When we then get to expressing it in code, I agree there's much to be said for erring toward explicit. Given that these issues are subtle, and that common constructions like `const` cover most of their appearance in typical code, the line between "implicit" and "magic" for less common constructions will be treacherous.

If we think of this as a layering of meta levels, then it seems to me so far that there are an infinite number of layers? It seems as though with successive `const` declarations we can repeatedly turn one layer's run time into the next layer's compile time?

Do we end up saying something like the following?

A macro or proc parameter can be marked `static` to indicate that the corresponding argument expression must be constant at the callsite. Every generic parameter is implicitly static. An expression is constant if it meets both of the following requirements:

* Every value used in the expression is either a literal or a constant defined textually ahead of the expression.
* Every proc used in the expression can be statically evaluated.

A proc can be statically evaluated if it meets all of the following requirements:

* It is defined textually ahead of the expression in which it is used.
* It is defined entirely in the Nim subset that is supported for compile-time evaluation.
* Every value used in its definition is a constant expression as defined above.

**Comment by Clyybber** [comment]
@zah Additionally the explicit approach is more clear to the programmer and people who want to read the code.
@deansher 
> If we think of this as a layering of meta levels, then it seems to me so far that there are an infinite number of layers? It seems as though with successive const declarations we can repeatedly turn one layer's run time into the next layer's compile time?
Exactly. AFAIK it doesn't enable any use cases a layering limited to 1 runtime and 1 compile time couldn't achieve too, but (as Araq stated on IRC) it allows for more freedom in the VM implementation, for example the VM could be implemented with a combination of the Nim Compiler and a fast C compiler, such as TinyCC, or even a C interpreter.

**Comment by timotheecour** [comment]
@zah 
> Sure, you can attach a value to the param symbol after the overload resolution is complete and you have selected the callee, but you need to be able to reason about static values earlier - while the overload resolution is underway.

Sorry, I'm still not seeing the issue, overload resolution would still work, for each `static` param, it'd just check whether the corresponding argument a `const` or evaluatable to `const` (eg literal or some proc calls).
The bugs with `static` (eg https://github.com/nim-lang/Nim/issues/10073) have to do with after overload resolution is done (inside the proc body with a static param), so would in theory be fixed by https://github.com/nim-lang/Nim/issues/10085#issuecomment-449839318 ; if you disagree, can you please provide a more detailed example where this would fail?

In other words, If I or someone started implementing https://github.com/nim-lang/Nim/issues/10085#issuecomment-449839318, what test case would fail?

**Comment by Araq** [comment]
Since then I made some progress and I think to understand the problem space better. There is no doubt that ``static T`` is adding tremendous power to Nim and plenty of code cannot be written without it. Having said that there are still issues with its *design* (ignoring the implementation complexity).

1. There is no clear type relation between ``static T`` and ``T``. It's true that ``var T`` has the same problem but overloading by ``var T`` is an inherently useful feature whereas overloading by ``static T`` is just an optimization that can also be achieved with a template that contains a ``when isLiteral(x)``.
2. ``static T`` is used for two different purposes:
- To turn values into types in generics.
- To enforce compiletime-evaluation in proc parameters.

3. Code like

```nim
proc p[T: static int](a: array[T, int]) = echo T
p([1, 2])
```

is confused. ``T`` is a type, ``echo`` could deal with ``typedesc`` for ``$`` too here so this could either echo "int" (!) or "2".

Much clearer would be 

```nim
proc p[T: static int](a: array[T, int]) = echo valueof(T)
p([1, 2])
```

An explicit ``valueof`` builtin to turn a ``static T`` into its value.

**Comment by deansher** [comment]
In `proc p[T: static int](...) ...` above, is @Araq saying

- `static int` is being used as a type constraint;
- `T` is a type that must meet that constraint; 
- and if `T` is inferred to be `2`, this treats `2` as a subtype of `static int`, in much the same way `type[int]` is a subtype of `type`?

**Comment by zah** [comment]
It's wrong to say that procs accepting static params are "just an optimisation". 

It's important to note that such procs become generic and their bodies may depend on the static parameter value (e.g. it may be used to determine the size of a local array, it may change the outcome of `when` statements and it may even affect overload resolution in certain cases). 

But even more important to me is the fact that once the proc is generic, the generic instantiation cache interacts with some other code generation mechanisms in the language such as the `{.global.}` pragma and static blocks guaranteed to be executed exactly once per instantiation (I've often called these "generic instantiation side effects" in previous discussions).

Since Nim doesn't have "generic" variables (e.g. `var foo[T] = ...`), generic procs using the `{.global.}` pragma are the way to simulate them. I'll give you a real world example where such variables are useful. In the [nim-chronicles](https://github.com/status-im/nim-chronicles) library, I allow the users to associate their log records with an arbitrary list of "topics" specified as strings. The topics are used for filtering and each topic has a current state in the program (required, disabled, filtered by log level, etc). I want the filtering code to be as short as possible and to avoid any table lookups, so I prefer the state of a particular topic to be kept in a global variable with a known static address. This is easily achieved using the following construct:

```nim
proc topicStateIMPL(topicName: static[string]): ptr TopicState=
  var topic {.global.} = TopicState(state: Normal, logLevel: NONE)
  var dummy {.global.} = registerTopic(topicName, addr(topic))
  return addr(topic)
```

Please note that this keeps the library ["open"](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) - the user may introduce new topic names without problems and Nim will ensure that exactly one variable is created for each unique topic name. The C compiler will be very happy to inline the generated short accessor function.



So, I wouldn't agree that `static` is somehow less useful than `var` and that it should be criticized while `var` is given a free pass.

--------------------

I don't see anything confusing about 

```nim
proc p[T: static int](a: array[T, int]) = echo T
p([1, 2])
```

Within the body of a proc, any static param can be considered as a constant of its respective type. Since it's perfectly legal to call `echo` with an int constant, the code should do just that and print 2 as a result.

The problematic case is where you don't explicitly specify T as a `static int` value. Only then, Nim's default array treatment will kick in, turning the generic parameter into a range type which may lead to some confusion.

---------------

I would argue that a future where `[1, 2, 3]` is recognized as a constant and typed as `static array[3, int]` is improvement over the status quo, because it will allow us to remove all on-demand `semConstExpr` calls present in overload resolution and the `is` operator handling.

**Comment by zah** [comment]
Ultimately, the signature of the type determines its internal representation.

If we had `type MyArray[N: static int; T] = ...`, then any instantiation will have the form: `MyArray[some_integer, some_type]`. Here, the `some_integer` value must be inserted as a son to the `tyGenericInst` type, so it will be naturally wrapped in `tyStatic` (which is exactly what the signature says after all). In other words, sigmatch will ensure that all expressions used to instantiate `MyArray` are fitted to the specific types in the signature. 

Now, let's see what happens before sigmatch has done its job. `tyStatic` will annotate expressions that were determined as known at compile time. In `tyStatic.n`, you hold the evaluated value without interfering with the original AST node. If most of the code in sem knows how to skip `tyStatic` just as it knows how to skip `tyVar`, I don't see any issues with such annotations. Perhaps my mistake was that I never pushed hard enough to add `tyStatic` to the skipTypes lists that include `tyVar`. Most of the places where `tyStatic` is a problem now should have existing special measures to deal with `tyVar`. For example, in the recent tuple construction issue, `tyVar` is just as unreasonable as `tyStatic` is for a field type of a tuple.

Once the `skipTypes` problems are taken care of, we'll still have the eager evaluation problems, but these are really about the generic instantiation process and not strictly about `static T`. Here is one example where `tyFromExpr` is required without involving any static params.

```nim
type Foo[T] = seq[type(bar(T))]
```

You don't know the type of `bar(T)` before you known what `T` is, so you must encode the whole `type(bar(T))` expression in a `tyFromExpr` placeholder.

**Comment by Araq** [comment]
> If most of the code in sem knows how to skip tyStatic just as it knows how to skip tyVar, I don't see any issues with such annotations. 

I'm not talking about the implementation. You didn't answer the question. Can static be nested or not?

Secondly, these annotations can do harm for ``macros.getType``.

> So, I wouldn't agree that static is somehow less useful than var and that it should be criticized while var is given a free pass.

I meant *overloading by static T*, especially:

```nim
proc p(x: int) = echo "int"
proc p(x: static int) = echo "static"

p 4
```

``4`` is of type ``int`` so it could be an ambiguity error or call the ``int`` variant instead.
